datalad/distribution/dataset.py

```diff
@@ -187,22 +189,61 @@ def path(self):
     @property
     def repo(self):
         """Get an instance of the version control system/repo for this dataset,
-        or None if there is none yet.
+        or None if there is none yet (or none anymore).
 
-        If creating an instance of GitRepo is guaranteed to be really cheap
-        this could also serve as a test whether a repo is present.
+        If testing the validity of an instance of GitRepo is guaranteed to be
+        really cheap this could also serve as a test whether a repo is present.
+
+        Note, that this property is evaluated every time it is used. If used
+        multiple times within a function it's probably a good idea to store its
+        value in a local variable and use this variable instead.
 
         Returns
         -------
-        GitRepo
+        GitRepo or AnnexRepo
         """
 
-        # Note: lazy loading was disabled, since this is provided by the
-        # flyweight pattern already and a possible invalidation of an existing
-        # instance has to be done therein.
-        # TODO: Still this is somewhat problematic. We can't invalidate strong
-        # references
-
+        # If we already got a *Repo instance, check whether it's still valid;
+        # Note, that this basically does part of the testing that would
+        # (implicitly) be done in the loop below again. So, there's still
+        # potential to speed up when we actually need to get a new instance
+        # (or none). But it's still faster for the vast majority of cases.
+        #
+        # TODO: Dig deeper into it and melt with new instance guessing. This
+        # should also involve to reduce redundancy of testing such things from
+        # within Flyweight.__call__, AnnexRepo.__init__ and GitRepo.__init__!
+        #
+        # Also note, that this could be forged into a single big condition, but
+        # that is hard to read and we should be well aware of the actual
+        # criteria here:
+        if self._repo is not None and realpath(self.path) == self._repo.path:
```

Comment:
shouldn't it be `_repo.realpath` then? wouldn't it make sense to store `.realpath` for the dataset as well to avoid realpathing upon each call?

Reply:
Re 1: _repo.path is realpath already. The actual constructor is only called by flyweight with already resolved path. `GitRepo.realpath` is not needed at all and can be removed, I think.

Re 2: Nope. If `Dataset.path` is a symlink, it could change during lifetime of the Dataset instance. See the tests I added in `test_dataset.py`. If we only resolve it once and store it, we would fail to notice that change.